internal/pkg/vm/internal/vmcontext/vmcontext.go

```diff
@@ -252,38 +252,80 @@ func (ctx *VMContext) Charge(cost types.GasUnits) error {
 
 var _ runtime.ExtendedInvocationContext = (*VMContext)(nil)
 
-var _ runtime.LegacyInvocationContext = (*VMContext)(nil)
-
-// LegacyVerifier returns an interface to the proof verification code
-func (ctx *VMContext) LegacyVerifier() verification.Verifier {
-	return &verification.RustVerifier{}
+func isBuiltinActor(code cid.Cid) bool {
+	return code.Equals(types.StorageMarketActorCodeCid) ||
+		code.Equals(types.InitActorCodeCid) ||
+		code.Equals(types.MinerActorCodeCid) ||
+		code.Equals(types.BootstrapMinerActorCodeCid) ||
+		code.Equals(types.PaymentBrokerActorCodeCid)
 }
 
-// LegacyMessage retrieves the message associated with this context.
-func (ctx *VMContext) LegacyMessage() *types.UnsignedMessage {
-	return ctx.message
+func isSingletonActor(code cid.Cid) bool {
+	return code.Equals(types.StorageMarketActorCodeCid) ||
+		code.Equals(types.InitActorCodeCid) ||
+		code.Equals(types.PaymentBrokerActorCodeCid)
 }
 
-// LegacyCreateNewActor creates an actor at the given address.
-// If the address is occupied by a non-empty actor, this method will fail.
-func (ctx *VMContext) LegacyCreateNewActor(addr address.Address, code cid.Cid) error {
+// CreateActor implemenets the ExtendedInvocationContext interface.
+func (ctx *VMContext) CreateActor(actorID types.Uint64, code cid.Cid, params []interface{}) address.Address {
+	if !isBuiltinActor(code) {
+		runtime.Abort("Can only create built-in actors.")
+	}
+
+	// Review: why are we checking this here? it assumes singleton actors always exist
```

Comment:
Yes, they are assumed to exist in the genesis state, not by the application of a message that could somehow invoke this method.